File: domain/src/main/java/com/ahn/domain/usecase/ToggleFavoriteCurrencyUseCase.kt
package com.ahn.domain.usecase

import com.ahn.domain.repository.FavoriteCurrencyRepository
import javax.inject.Inject

class ToggleFavoriteCurrencyUseCase @Inject constructor(
    private val repository: FavoriteCurrencyRepository
) {
    suspend operator fun invoke(currencyCode: String) {
        if (repository.isFavorite(currencyCode)) {
            repository.removeFavorite(currencyCode)
        } else {
            repository.addFavorite(currencyCode)
        }
    }
}

File: ios/Packages/Domain/Sources/Domain/Favorite/UseCase/ToggleFavoriteCurrencyUseCase.swift
public struct ToggleFavoriteCurrencyUseCase: Sendable {
    private let repository: any FavoriteCurrencyRepository

    public init(repository: any FavoriteCurrencyRepository) {
        self.repository = repository
    }

    public func callAsFunction(_ currencyCode: String) async throws {
        if try await repository.isFavorite(currencyCode) {
            try await repository.removeFavorite(currencyCode)
        } else {
            try await repository.addFavorite(currencyCode)
        }
    }
}
